### PR TITLE
Re-add type icons (again)!

### LINKED ISSUE #19318: New inspector doesn't show type icons
**Godot version:**
 8684b63

**Issue description:**
The new inspector doesn't show the type icons when enabled.

### PR DESCRIPTION
**Pull request by YeldhamDev**
Sorry Juan, but I'm not letting these die so easily. Fixes #19318.
screenshot at 2018-09-09 21-50-41

### CODE REVIEW

**Comment by groud** [comment]
I think "On" is more user-friendly than "True" from a user perspective. 
But maybe "Active" would have been even better.

**Comment by akien-mga** [comment]
I think "On" is better than "True" - think of translations, if I see a widget with "Vrai" in French, I'll have no clue what it means. The French translation for "On" is actually "Activé" ("Enabled").

At any rate, this should be discussion in another issue/PR as it's not related to type icons, and should not be part of its commit.

**Comment by YeldhamDev** [comment]
@akien-mga Good point, back to "On" for now.

File: editor/editor_inspector.cpp [hunk 1]
```diff
@@ -218,11 +222,16 @@ void EditorProperty::_notification(int p_what) {
...
-			ofs += checkbox->get_width();
+			ofs += checkbox->get_width() + get_constant("hseparator", "Tree");
 		} else {
 			check_rect = Rect2();
 		}
+		if (EDITOR_DEF("interface/editor/show_type_icons", false)) {
+			if (!type_icon.is_null())
+				draw_texture(type_icon, Vector2(ofs, (size.height / 2) - (type_icon->get_height() / 2)));
+
+			ofs += 20; // Hardcoded value, so there's still a space if there's no icon for the property.
```
Comment (akien-mga) [inline editor/editor_inspector.cpp]: How does it work with HiDPI and e.g. 200% upscaled icons?

**Comment by capnm** [comment]
I agree here with Juan, it just clutters the GUI, the type is already clear from the right side.

**Comment by groud** [comment]
> I agree here with Juan, it just clutters the GUI, the type is already clear from the right side.

This has been discussed several time. I don't like it myself but a lot of people do. As this is optional it's not a problem to have it back.

**Comment by capnm** [comment]
> This has been discussed several time. I don't like it myself but a lot of people do. As this is optional it's not a problem to have it back.

Okay, if I can always turn it off, then make Godot great again ;-)

**Comment by groud** [comment]
> Okay, if I can always turn it off, then make Godot great again ;)

Don't worry, it is not even on by default :)

File: editor/editor_inspector.cpp (hunk 1 above)
Comment (YeldhamDev) [reply editor/editor_inspector.cpp]: Yep, I'll multiply that by EDSCALE right now.

**Comment by williamd1k0** [comment]
I was missing them -😥

**Comment by akien-mga** [comment]
For the reference, I haven't merged this yet because I'm not sure it's implemented in the most efficient way. Currently it will allocate and set type icons for all property editors, and call an `update()` each time, even if the option for type icons is not enabled. That seems quite wasteful of resources IMO.

**Comment by YeldhamDev** [comment]
@akien-mga `set_type_icon()` now only calls `update()` when type icons are enabled.

And for the editors setting the icons themselves, I really don't see a better approach to this. It let's the `EditorProperty` itself decide which icon is best suited for its type.

**Comment by akien-mga** [comment]
> `set_type_icon()` now only calls `update()` when type icons are enabled.

That's better for the no-type icons scenario indeed, but it's still a lot of `update()` calls happening in the same frame for users of type icons, no? (maybe I misunderstand when those will be called though).
If they do all happen at the same time (at least for a given Node's properties in the inspector), maybe `call_deferred("update")` could improve things, as I think `MessageQueue` will batch identical function calls (needs confirmation though). Alternatively there could be a dirty/needs_update boolean that could be handled in the next refresh.

**Comment by YeldhamDev** [comment]
The `update()` call happens on:
- Tree entering.
- Editor settings change (to check if it was enabled/disabled).
- Theme change (update icon colors).

I however made that when the editor settings are changed, to check if the `show_type_icons` value has actually changed before updating. Avoiding calling it on any change.

I don't think there's a need to defer the call, `set_label()` calls for example, happen all at the same time, without deferring.

**Comment by Norrox** [comment]
type icons does not work in godot 3.1 beta 2
fyi ;)

**Comment by Anutrix** [comment]
Note that the "Show Type Icons" is still present Interface section in Editor Settings but does nothing.
Also, checking or unchecking does not show the "Reset to default value" button that usually shows up to the left of any non-default values of settings.
Just wanted to mention this. Is it an unrelated bug? Should I create a separate issue?

**Comment by Calinou** [comment]
> Note that the "Show Type Icons" is still present Interface section in Editor Settings but does nothing.
Also, checking or unchecking does not show the "Reset to default value" button that usually shows up to the left of any non-default values of settings.

Maybe this setting was left over from an old configuration file? See https://github.com/godotengine/godot/issues/22099.

**Comment by Anutrix** [comment]
> 
> 
> > Note that the "Show Type Icons" is still present Interface section in Editor Settings but does nothing.
> > Also, checking or unchecking does not show the "Reset to default value" button that usually shows up to the left of any non-default values of settings.
> 
> Maybe this setting was left over in your local configuration as it came from an old configuration file? See #22099. (I can't see any instances of `show_type_icons` in the current codebase, and I don't see it in my configuration file.)

Not finding in code base confused me too. I didn't know the list of editor settings was generated from the config files. I thought it was generated/hard-coded in the code base somewhere. Thx for the info.

**Comment by YeldhamDev** [comment]
By the fact that this isn't going anywhere, and me growing tired of dealing with the conflicts, I'll be closing this.

**Comment by Xrayez** [comment]
> If they do all happen at the same time (at least for a given Node's properties in the inspector), maybe call_deferred("update") could improve things, as I think MessageQueue will batch identical function calls (needs confirmation though)

Doesn't seem like so, I've seen on numerous occasions how it's optimized with `_queue_update` or `_make_dirty` methods even with `call_deferred` used.

https://github.com/godotengine/godot/blob/24e1039eb6fe32115e8d1a62a84965e9be19a2ed/scene/scene_string_names.cpp#L171

https://github.com/godotengine/godot/blob/24e1039eb6fe32115e8d1a62a84965e9be19a2ed/scene/resources/visual_shader.cpp#L1294-L1301

**Pull request closed without merging**